cmd/snap-confine/seccomp-support.c

```diff
@@ -198,8 +211,15 @@ int sc_apply_seccomp_bpf(const char *filter_profile)
 		.len = num_read / sizeof(struct sock_filter),
 		.filter = (struct sock_filter *)bpf,
 	};
-	if (prctl(PR_SET_SECCOMP, SECCOMP_MODE_FILTER, &prog) != 0) {
-		die("cannot apply seccomp profile");
+	if (seccomp(SECCOMP_SET_MODE_FILTER, SECCOMP_FILTER_FLAG_LOG, &prog)) {
+		if (errno == ENOSYS)
+			debug("kernel doesn't support the seccomp(2) syscall");
+		else if (errno == EINVAL)
+			debug("kernel may not support the SECCOMP_FILTER_FLAG_LOG flag");
+
```

Comment:
This seems to be where the intended future check could be better (ie, using 'may'). What I was hoping for was that if SECCOMP_FILTER_FLAG_LOG is not available, then snapd would log that seccomp is mediating with EPERM but without logging so people understand the behavior change. Before, commands run under strict mode confinement with access violations would be killed with logging, but now on systems without SECCOMP_FILTER_FLAG_LOG support, they will be returned EPERM and continue to run with no logging, which may be confusing since the new denied behavior looks more like the previous allowed behavior than the previous denied behavior.

All that said, logging the fallback behavior IMO should happen at the snapd level once per snapd invocation since logging in snap-seccomp or in snap-confine would be way too chatty. @mvo5 - I suspect this would be more for whoever picks this up than for Tyler.

Reply:
That's a good point that logging the fallback shouldn't happen in `snap-seccomp` as it would be spammy in the logs.

libseccomp upstream was against exporting a function specifically for checking the validity of a filter flag or action. Applications linking against libseccomp will have the option to check the validity of the `SECCOMP_FILTER_FLAG_LOG` flag indirectly but using an "API level" based check that is currently being implemented. However, that check lumps together multiple features and, in this case, a single API level will be used to represent `SECCOMP_FILTER_FLAG_LOG`, `SECCOMP_RET_LOG`, and the unrelated `SECCOMP_RET_KILL_PROCESS`. That means that the kernel, libseccomp, and libseccomp-golang will all have to be in sync for libseccomp to report that the API level is supported.

Therefore, I think snapd would want to directly use the `seccomp(2)` syscall for checking if a filter flag is supported. Snapd could just use the technique that libseccomp is going to be using behind the scenes. I added a kernel selftest for the technique so that we can be sure that the kernel doesn't violate the agreement with userspace about how to check if a filter flag is supported. You can see the kernel selftest change [here](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=2b7ea5b5b5799f2878ed454bb48032bed6d101d3). Calling `seccomp(SECCOMP_SET_MODE_FILTER, SECCOMP_FILTER_FLAG_LOG, NULL)` should return -1 with errno set to `EFAULT` if the flag is supported by the kernel. Other errno values mean other things such as `EINVAL` meaning that the flag is not supported and `ENOSYS` meaning that the `seccomp(2)` syscall isn't even supported (so there's no way to even specify a filter flag).